### PR TITLE
Adds the ability to write certs as a pem bundle or individual files

### PR DESCRIPTION
**Pull request by k-s-dean**

### CODE REVIEW

File: roles/vault_pki/README.md [hunk 1]
```diff
@@ -47,4 +47,5 @@ Role variables
     * `vault_pki_write_certificates_host:` The host on which certificates will be written to. (default: "localhost")
     * `vault_pki_certificates_directory`: directory to output certificate files to.
     * `vault_pki_write_certificates`: whether to write generated certificates to a file or not (default: `false`)
```
Comment (markgoddard) [inline roles/vault_pki/README.md:49]: unrelated: I notice this variable isn't used

File: roles/vault_pki/defaults/main.yml [hunk 2]
```diff
@@ -23,6 +23,7 @@ vault_pki_intermediate_key_bits: 4096
 vault_pki_intermediate_roles: {}
 
 vault_pki_generate_certificates: false
+vault_pki_write_pem_bundle: false
 
 vault_pki_write_certificates_host: "localhost"
 vault_pki_write_certificates_directory: ""
```
Comment (markgoddard) [inline roles/vault_pki/defaults/main.yml:29]: unrelated: This isn't referenced anywhere else, but `vault_pki_certificates_directory` is.

File: roles/vault_pki/README.md [hunk 3]
```diff
@@ -47,4 +47,5 @@ Role variables
...
+    * `vault_pki_write_pem_bundle`: output the generated certificate and private key into a single file (default: `false`)
```
Comment (markgoddard) [inline roles/vault_pki/README.md:50]: Default of false is a change in behaviour. Is that what we want?

File: roles/vault_pki/tasks/main.yml [hunk 4]
```diff
@@ -17,5 +17,6 @@
 - name: "Generate Certificates"
   include_tasks: "create_cert.yml"
   when:
+    # CI
```
Comment (markgoddard) [inline roles/vault_pki/tasks/main.yml:20]: CI?

File: tests/test_vault.yml [hunk 5]
```diff
@@ -61,6 +64,18 @@
               ip_sans: "127.0.0.1"
               alt_names: "example.com"
               exclude_cn_from_sans: true
+    - name: Configure PKI
```
Comment (markgoddard) [inline tests/test_vault.yml:67]: You're hitting an ansible quirk where multiple import_role tasks for the same role in one play will have the task variables from the last import overriding earlier imports. The solution is to use include_role instead, which does a dynamic include and works as expected.

Comment (markgoddard) [reply tests/test_vault.yml:67]: The name of the task should be changed to reflect the difference between the two invocations of the role.

File: tests/test_vault.yml [hunk 6]
```diff
@@ -61,6 +64,18 @@
...
+        name: vault_pki
+      vars:
+        vault_token: "{{ vault_keys.root_token }}"
+        vault_pki_root_ca_name: "OS-TLS-ROOT"
+        vault_pki_intermediate_ca_name: "OS-TLS-INT"
+        vault_pki_root_create: False
+        vault_pki_intermediate_create: False
+        vault_pki_generate_certificates: True
+        vault_pki_write_pem_bundle: True
+        vault_pki_certificate_subject:
```
Comment (markgoddard) [inline tests/test_vault.yml:78]: It's probably time to verify that this has actually done what we wanted it to. Bare minimum would be checking for existence of the generated certs, ideally checking the trust chain too.

**Review by markgoddard — CHANGES_REQUESTED** [review]

File: tests/test_vault.yml [hunk 7]
```diff
@@ -61,6 +64,18 @@
...
           - role: 'ServerCert'
```
Comment (markgoddard) [inline tests/test_vault.yml:79]: Would we ever want to generate some as PEM bundles and some as separate cert & key in a single run?

File: .github/workflows/pull_request.yml [hunk 8]
```diff
@@ -65,11 +65,11 @@ jobs:
           else
               ansible_package=ansible-core
           fi
-          python3 -m pip install $ansible_package==$ansible_version.* docker git+https://github.com/stackhpc/ansible-modules-hashivault@stackhpc
+          python3 -m pip install $ansible_package==$ansible_version.* docker git+https://github.com/stackhpc/ansible-modules-hashivault@stackhpc-test
```
Comment (mnasiadka) [inline .github/workflows/pull_request.yml:68]: TODO: sort out proper management of that fork

File: roles/vault_pki/README.md (hunk 1 above)
Comment (mnasiadka) [reply roles/vault_pki/README.md:49]: Removed in a different PR

File: tests/test_vault.yml (hunk 6 above)
Comment (mnasiadka) [reply tests/test_vault.yml:78]: TODO

File: tests/test_vault.yml (hunk 7 above)
Comment (mnasiadka) [reply tests/test_vault.yml:79]: Probably makes sense, will rework.

File: roles/vault_pki/defaults/main.yml (hunk 2 above)
Comment (k-s-dean) [reply roles/vault_pki/defaults/main.yml:29]: Thanks. fixed.

File: tests/test_vault.yml (hunk 7 above)
Comment (k-s-dean) [reply tests/test_vault.yml:84]: I don't think it needs reworked. 
I would consider whats done in the test_vault.yml as a single run ?

File: roles/vault_pki/tasks/main.yml (hunk 4 above)
Comment (k-s-dean) [reply roles/vault_pki/tasks/main.yml]: Removed

File: tests/test_vault.yml (hunk 5 above)
Comment (k-s-dean) [reply tests/test_vault.yml]: Done.

File: roles/vault_pki/README.md (hunk 3 above)
Comment (k-s-dean) [reply roles/vault_pki/README.md]: Most things don't use pem bundles. They use a cert and key. So I'd say yes we want this change in behaviour.

**Review by markgoddard — CHANGES_REQUESTED** [review]

File: roles/vault_pki/README.md [hunk 9]
```diff
@@ -46,4 +46,5 @@ Role variables
     * `vault_pki_generate_certificates`: whether to generate leaf certificates or not (default: `false`)
```
[the remark this answers is not in the document]
Comment (mnasiadka) [reply roles/vault_pki/README.md:46]: Done.

File: roles/vault_pki/tasks/create_cert.yml [hunk 10]
```diff
@@ -9,18 +20,63 @@
     role: "{{ item.role }}"
     extra_params: "{{ item.extra_params }}"
   loop: "{{ vault_pki_certificate_subject }}"
-  register: certificate_data
+  loop_control:
+    label: "{{ item.common_name }}"
+  register: cert_data
```
[the remark this answers is not in the document]
Comment (mnasiadka) [reply roles/vault_pki/tasks/create_cert.yml:25]: It looks to be idempotent, but we can skip that.

File: roles/vault_pki/tasks/create_cert.yml [hunk 11]
```diff
@@ -1,5 +1,16 @@
 ---
+- name: "Check if Cert already exists in certificates directory"
+  stat:
+    path: "{{ vault_pki_write_certificates_directory }}/{{ item.common_name | replace(' ', '-') }}.crt"
+  register: cert_stat_file
+  delegate_to: "{{ vault_pki_write_certificates_host }}"
+  loop: "{{ vault_pki_certificate_subject }}"
+  loop_control:
+    label: "{{ item.common_name }}"
```
[the remark this answers is not in the document]
Comment (mnasiadka) [reply roles/vault_pki/tasks/create_cert.yml:9]: Done.

File: roles/vault_pki/tasks/prechecks.yml [hunk 12]
```diff
@@ -3,7 +3,7 @@
   fail:
     msg: "variable {{ item.name }} is not set"
   when:
-    - vars[item.name] | length == 0
+    - vars[item.name] | default() | length == 0
```
[the remark this answers is not in the document]
Comment (mnasiadka) [reply roles/vault_pki/tasks/prechecks.yml:6]: It did help, but let me try with is defined.

File: tests/test_vault.yml [hunk 13]
```diff
@@ -59,9 +66,52 @@
...
+        executable: /bin/bash
+      become: true
+
+    - name: verify certificate chain
+      command: | 
+        openssl verify -CAfile /tmp/CA-CHAIN.pem
+        /tmp/{{ item }}
+      register: verify_result
+      failed_when: verify_result.rc >= 1
+      become: true
```
[the remark this answers is not in the document]
Comment (mnasiadka) [reply tests/test_vault.yml:113]: dropped, let's see if that works.

File: roles/vault_pki/tasks/create_cert.yml (hunk 10 above)
[the remark this answers is not in the document]
Comment (markgoddard) [reply roles/vault_pki/tasks/create_cert.yml:26]: if it's idempotent then it's not really regenerating certificates is it?

[the remark this answers is not in the document]
Comment (mnasiadka) [reply roles/vault_pki/tasks/create_cert.yml:26]: Yup, it's idempotent, checked to be sure.

File: roles/vault_pki/defaults/main.yml (hunk 2 above)
Comment (mnasiadka) [reply roles/vault_pki/defaults/main.yml]: Done.

File: roles/vault_pki/tasks/create_cert.yml [hunk 14]
```diff
@@ -1,5 +1,17 @@
...
+  when: vault_pki_write_certificate_files | bool
+
 - name: "Generate Certificate"
+  vars:
+    cert_file: "{{ cert_stat_file.results | selectattr('item.common_name', 'match', item.common_name) | first }}"
```
Comment (markgoddard) [inline roles/vault_pki/tasks/create_cert.yml:14]: Is it used?

Comment (mnasiadka) [reply roles/vault_pki/tasks/create_cert.yml]: No, removed.

**Review by markgoddard — APPROVED** [review]

**Pull request merged**